### PR TITLE
Feature request: Scripting hook

### PR DESCRIPTION
**Pull request by umireon**
As the author of the EditorConfig plugin for CotEditor,
I want scripting hooks such as onload and onsave to be integrated in CotEditor,
To make the plugin works seamlessly (ie without additional user interaction).

Script bundles have `Contents/Info.plist`, which can contain additional meta data for the bundle.
Given the following property list:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>CotEditorHandlers</key>
  <array>
    <string>document opened</string>
    <string>document saved</string>
  </array>
</dict>
</plist>
```

CotEditor can configure the scripting hook only by reading Info.plist.
Note that a property list can be parsed easily with `NSDictionary`.

Finally, Event delivery with AppleScript can be implemented with Apple Event.
Example: [Messages.app](https://discussions.apple.com/thread/6476545?tstart=0)

The basic concept of this proposal is based on the discussion in #513.
The following points is also carefully considered:
* No startup delay: Hook registration can be performed without any external code execution
* No user configuration: Hook configuration is contained in plugin itself
* Compatible with AppleScript and JXA: Hook is written in the form of handler

## Changeset Description
Add hook mechanics for AppleScript plugins

Script bundles can have `Info.plist`, which is useful to store metadata.
For example, the following key-value set can be used to determine
whether the event of document opened should delivered to the plugin:

```
<key>CotEditorHandlers</key>
<array>
<string>document opened</string>
</array>
```

By this change, the `CotEditorHandlers` in `Info.plist` of script bundle
lugins will be read and registered to `ScriptManager`.

Additionally, the API to invoke the registered handlers was introduced:
  `ScriptManager.dispatch(_:toHandlersAt)`

The following scripting events added:
* document opened
* document saved

### CODE REVIEW

**Comment by umireon** [comment]
TL;DR in Japanese: Apple Eventを活用すると、ユーザースクリプトへのフックを簡潔なコードで導入できますから、ぜひ取り入れてください。よろしくおねがいします。

**Comment by umireon** [comment]
Works as this (plz ignore being silly plugin; just an example)
screencast 2016-12-04 00-33-41 mp4

File: CotEditor/Sources/Document.swift [hunk 1]
```diff
@@ -337,6 +337,8 @@ final class Document: NSDocument, AdditionalDocumentPreparing, EncodingHolder {
             ?? Defaults[.syntaxStyle]
         
         self.setSyntaxStyle(name: styleName)
+        
+        ScriptManager.shared.dispatchEvent(documentOpened: self)
```
Comment (1024jp) [inline CotEditor/Sources/Document.swift]: `read(from:, ofType:)` is invoked not only on file opening but also on file reverting (namely, both on "Revert To..." action and on resolving a file update conflict). 

Thus, I suggest not to call `ScriptManager.shared.dispatchEvent(documentOpened: self)` here but somewhere else. Well..., I'd say `didMakeDocumentForExisitingFile(url:)` is better. The method is invoked after `read(from:, ofType:)` and before document-related views are loaded.

File: CotEditor/Sources/ScriptManager.swift [hunk 2]
```diff
@@ -204,6 +209,64 @@ final class ScriptManager: NSObject, NSFilePresenter {
...
+    /// - parameters:
+    ///   - document: the document to dispatch an Apple event
+    ///   - eventID: the event ID to be set in the returned event
+    ///
+    /// - returns: a descriptor for an Apple event by the `Document`
+    func createEvent(by document: Document, eventID: AEEventID) -> NSAppleEventDescriptor {
+        let event = NSAppleEventDescriptor(eventClass: AEEventClass(code: "cEd1"), eventID: eventID, targetDescriptor: nil, returnID: AEReturnID(kAutoGenerateReturnID), transactionID: AETransactionID(kAnyTransactionID))
+        
+        //
+        let documentDescriptor = document.objectSpecifier.descriptor!
```
Comment (1024jp) [inline CotEditor/Sources/ScriptManager.swift]: I wanna just make sure; is it safe to force unwrap?

File: CotEditor/Sources/Document.swift (hunk 1 above)
Comment (umireon) [inline CotEditor/Sources/Document.swift]: Nice, i did not sure about that.
I also suggest `document resumed` event but not in this PR 😄

File: CotEditor/Sources/ScriptManager.swift (hunk 2 above)
Comment (umireon) [inline CotEditor/Sources/ScriptManager.swift]: It is definitely not safe because of [the poor documentation by Apple](https://developer.apple.com/reference/foundation/nsscriptobjectspecifier/1410018-descriptor).

**Comment by umireon** [comment]
I will work on the help in a separate PR #650

**Comment by umireon** [comment]
If `document.objectSpecifier.descriptor` is nil, the following error message (`Can’t get file of "BUG: document.objectSpecifier.descriptor was nil"`) will be delivered to the user script which is written in the AppleScript.

console 2016-12-04 16-17-39

```applescript
using terms from application "CotEditor"
	on document saved theDocument
		set thePath to file of theDocument
	end document saved
end using terms from
```

As for JXA, `typeof document === "string"` when the error occurred, but it causes various errors because of its loose property handling 😢 .
```javascript
CotEditor = Application.currentApplication()
CotEditor.includeStandardAdditions = true
function documentSaved(document) {
	CotEditor.displayAlert("Saved " + document.file().toString())
}
```

**Good news** `document.objectSpecifier.descriptor` will never be `nil` because the descriptor is well-defined in `CotEditor.sdef`.

File: CotEditor/Sources/ScriptManager.swift [hunk 3]
```diff
@@ -254,6 +321,20 @@ final class ScriptManager: NSObject, NSFilePresenter {
     }
     
     
+    private func loadScriptInfo(at url: URL) {
```
Comment (umireon) [inline CotEditor/Sources/ScriptManager.swift]: The `bug` section on documentation is added.
IMO that condition will never be happen without macOS's bug while sdef is sane.

Comment (umireon) [inline CotEditor/Sources/ScriptManager.swift]: I quitted to pass `nil` as the direct parameter because `nil` is regareded as the current application in the user script and that fact causes strange behaviors.

**Comment by 1024jp** [comment]
Thanks! now it's fine enough to merge :D

**Comment by 1024jp** [comment]
I added your name to Acknowledgements.html files.

If you wanna add/change your name, URL, etc., feel free to create a new PR.

**Comment by kour1er** [comment]
Sorry about bringing this old issue up, but I'm very interested in the hook mentioned here but I can't seem to get it to work. I've tried with the template [https://github.com/coteditor/SampleScripts/pull/1](url) but no luck. I just wanted to check that it's still a supported feature and I'm probably missing something obvious? Thanks for any pointers
